aten/src/ATen/FunctionalTensorWrapper.cpp

```diff
@@ -183,6 +183,10 @@ void FunctionalTensorWrapper::replace_(const Tensor& other) {
   // out= ops are allowed to resize the output tensors, mutating both the data and metadata of the tensor.
   // We need to propagate that metadata mutation to the wrapper (new size).
   set_sizes_and_strides(value_.sizes(), value_.strides());
+  set_storage_offset(value_.storage_offset());
+  if (dtype() != value_.unsafeGetTensorImpl()->dtype() || layout() != value_.unsafeGetTensorImpl()->layout()) {
+    value_ = value_.to(c10::TensorOptions().dtype(dtype()).layout(layout()));
+  }
```

Comment:
Right now I do this by making assumptions about which classes of metadata can be mutated by an operator
- size/stride/storage offset can all be modified by an inplace op (e.g. out= ops that resize, or transpose_())
- dtype/layout are never modified by an operator, which lets me assume that the dtype/layout on the wrapper are "correct", and that I can propagate them to the inner tensor if they diverge

Reply:
With this change (and some more further in the stack, I was able to get all of the LTC tests to pass, which makes me reasonable convinced that I haven't missed any weird cases.

Now that I think about it though, the absolutely correct thing to do would probably just to run the original operator with meta tensors, and compare the output meta tensor's metadata to the functional op's output.

We can't actually do that though without full meta tensor coverage.

Reply:
Do you need to do this with memory format too?

Reply:
Yes doing this with memory format will be good. 
Another thought (that's for all functionalized inplace ops) - if `a` is broadcasted, inplace op will balk, but `copy_to` will happily proceed.

Reply:
Is there... a good way to check if two tensors have the same memory format? (what if they're both channels last, but one is contiguous and the other isn't?).

I guess I was thinking something like this, which looks pretty awful:
```
auto other_chan_last = other.unsafeGetTensorImpl()->is_strides_like_channels_last();
auto same_channels_last = is_strides_like_channels_last() == other_chan_last;

auto other_chan_last_3d = other.unsafeGetTensorImpl()->is_strides_like_channels_last_3d();
auto same_channels_last_3d = is_strides_like_channels_last_3d() == other_chan_last_3d

if (!same_channels_last || !same_channels_last_3d) {
    auto mem_format = other_chan_last ? MemoryFormat::ChannelsLast : other_chan_last_3d ? MemoryFormat::ChannelsLast3d : MemoryFormat::Preserve;
     value_ = value_.to(c10::TensorOptions().dtype(dtype()).layout(layout()), mem_format)
}

```

Reply:
Do they have the same sizes? One way is to compare strides